source/binding/Bitstream.cpp

```diff
@@ -0,0 +1,656 @@
+    if (unpackWidth) {
+        if (unpackWidth < totalWidth) { // left-aligned so trim rightmost
+            auto trimWidth = totalWidth - unpackWidth;
+            while (trimWidth >= rightWidth) {
+                rightIndex--;
+                trimWidth -= rightWidth;
+                rightWidth = getWidth(*packed[rightIndex]);
+            }
+            rightWidth -= static_cast<bitwidth_t>(trimWidth);
+        }
+        // For unpack, extraBits is for the first block
+        // For pack, extraBits is for the last block
+        extraBits = unpackWidth % sliceSize;
+    }
+
+    std::vector<ConstantValue> result;
+    result.reserve(std::max(packed.size(), numBlocks));
+    auto sliceOrAppend = [&](PackIterator iter) {
+        if (rightWidth == getWidth(**iter))
+            result.emplace_back(std::move(**iter));
+        else {
+            bitwidth_t bit = 0;
+            result.emplace_back(slicePacked(iter, packed.cend(), bit, rightWidth));
+        }
+    };
+
+    while (numBlocks > 1) {
+        auto index = rightIndex;
+        auto width = rightWidth;
+        auto slice = sliceSize;
+        if (extraBits) {
+            slice = extraBits;
+            extraBits = 0;
+        }
+        while (slice >= width) {
+            index--;
+            slice -= width;
+            width = getWidth(*packed[index]);
+        }
+
+        // A block composed of bits from the last "slice" bits of packed[index] to the first
+        // "rightWidth" bits of packed[rightIndex]
+        auto iter = packed.cbegin() + index;
+        if (slice) {
+            auto bit = static_cast<bitwidth_t>(width - slice);
+            result.emplace_back(
+                slicePacked(iter, packed.cend(), bit, static_cast<bitwidth_t>(slice)));
+            width -= static_cast<bitwidth_t>(slice);
+        }
+        iter++;
+        auto nextIndex = index;
+        while (++index < rightIndex)
+            result.emplace_back(std::move(**iter++));
+        if (index == rightIndex)
+            sliceOrAppend(iter);
+        rightIndex = nextIndex;
+        rightWidth = width;
+        numBlocks--;
+    }
+
+    // For pack, the last block may be smaller than slice size
+    auto iter = packed.cbegin();
+    for (size_t i = 0; i < rightIndex; i++)
+        result.emplace_back(std::move(**iter++));
+    sliceOrAppend(iter);
+
+    return result;
+}
+
+/// Performs unpack operation of streaming concatenation target on a bit-stream.
+static bool unpackConcatenation(const StreamingConcatenationExpression& lhs, PackIterator& iter,
+                                const PackIterator iterEnd, bitwidth_t& bitOffset,
+                                size_t& dynamicSize, EvalContext& context,
+                                SmallVector<ConstantValue>* dryRun = nullptr) {
+    for (auto stream : lhs.streams()) {
+        if (stream->kind == ExpressionKind::Streaming) {
+            auto concat = stream->as<StreamingConcatenationExpression>();
```

Comment:
This should be auto&. It's unfortunate that this compiled, making a copy here is not good.

Reply:
Yes, I change it for all places. Usually clang should warn about this, I don't know why it doesn't in this case.
My original code referenced ConcatenationExpression::operands(). The existing code has a lot of
```c++
        for (auto operand : operands()) {
```
I assume I should change them too.

Reply:
I was referring to the `auto concat = stream->as<>`, the as() call returns a reference right? So `concat` will be a copy of the object, unless I'm missing something.

Reply:
Yes, you are right. I will correct it.

Reply:
I found another line having the same problem and corrected it too.
```c++
        auto& source = rhs.as<StreamingConcatenationExpression>();
```